MSC3916: Authentication for media (#3916)
matrix-org/matrix-spec-proposals: proposals/3916-authentication-for-media.md

Author: Since this is only useful in combination with MSC3911, it should not complete FCP until MSC3911 also completes, otherwise we'll end up with spec changes with no useful usecase.

Author: oh, this isn't true. Authenticated media helps us fix https://github.com/matrix-org/synapse/issues/2133.

Reviewer: yea, I think this MSC stands well on its own merits, without support from other MSCs.

MSC3911 depends on this MSC, but this MSC does not depend on MSC3911 to be useful.

Reviewer: Reviewers: Please compare this MSC against https://github.com/matrix-org/matrix-spec-proposals/pull/2461 , which is proposed for *rejection/closure*.

Reviewer: it's still useful to compare the MSCs, given the ideas and concepts of 2461 are meant to be incorporated here.

  store. [synapse#2133](https://github.com/matrix-org/synapse/issues/2133)
* When a media event is redacted, the media it used remains visible to all.
  [synapse#1263](https://github.com/matrix-org/synapse/issues/1263)
* There is currently no way to delete
  media. [matrix-spec#226](https://github.com/matrix-org/matrix-spec/issues/226)
* If a user requests GDPR erasure, their media remains visible to all.
* When all users leave a room, their media is not deleted from the server.

These problems are all difficult to address currently, because access to media
is entirely unauthenticated. The first step for a solution is to require user
authentication. Once that is done, it will be possible to impose authorization
requirements to address the problems mentioned above. (See, for example,
[MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911) which
builds on top of this MSC.)

This proposal supersedes [MSC1902](https://github.com/matrix-org/matrix-spec-proposals/pull/1902).

## Proposal

1. New endpoints

   The existing `/_matrix/media/v3/` endpoints are to be deprecated, and replaced
   by new endpoints under the `/_matrix/client` and `/_matrix/federation`

>    hierarchies.

Reviewer: per https://github.com/matrix-org/matrix-spec-proposals/pull/3860/files#r1005176480 , `allow_redirect` should be dropped from the client endpoints and always expect redirection. Preferably on the federation endpoints too, but see https://github.com/matrix-org/matrix-spec-proposals/pull/3916/files#r1320697665

Reviewer: (dropped)

   The current
   [`/download`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaid)
   and
   [`/thumbnail`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)
   endpoints take an `allow_remote` query parameter, indicating whether the
   server should request remote media from other servers. This is redundant
   with the new endpoints, so will not be supported.

   Servers should never return remote media from `GET
   /_matrix/federation/v1/media/download` or `GET
   /_matrix/federation/v1/media/thumbnail`; indeed, the `serverName` is
   included in the URI only for consistency with the CS-API.

   `/_matrix/client/v1/media/download` and
   `/_matrix/client/v1/media/thumbnail` should return remote media as normal.
   
3. Authentication on all endpoints

   Currently, the `/download` and `/thumbnail` endpoints have no authentication
   requirements. Under this proposal, the new endpoints will be authenticated
   the same way as other endpoints: they will require an `Authorization` header
   which must be `Bearer {accessToken}` for `/_matrix/client`, or the signature

>    for `/_matrix/federation`.

Reviewer: This fixes https://github.com/matrix-org/matrix-spec/issues/313

>    for `/_matrix/federation`.

Reviewer 2: This is incompatible with eg. hosting a static directory (or unauthenticated s3/b2 bucket) for files, eg. emojis (eg. `/download/my.server.xyz/blobfox_snuggle.png`). Are there any considerations that could be taken for this?

Reviewer: see https://github.com/matrix-org/matrix-spec-proposals/pull/3916#discussion_r1320697665

4. Updated response format

   * For the new `/_matrix/client` endpoints, the response format is the same as
     the corresponding original endpoints.

   * To enable future expansion, for the new `/_matrix/federation` endpoints,
     the response is

>      [`multipart/mixed`](https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html)

Reviewer: `multipart/mixed` is deprecated. Long live `multipart/form-data` - https://www.rfc-editor.org/rfc/rfc7578

Form data might be better as a response type anyways, as it allows us to canonically reference the JSON body with a `name`.

Author: TIL. https://www.rfc-editor.org/rfc/rfc7578#section-4.3 appears to be the relevant section here. It's not *entirely* clear to me that `multipart/mixed` is deprecated as a whole, but I'm happy to take suggestions.

Reviewer: I think given the implementation complexity, it's not really a concern either way imo. We can continue to use `multipart/mixed` with no significant issues.

Reviewer: (resolving - no action)

     content with two parts: the first must be a JSON object (and should have a
     `Content-type: application/json` header), and the second is the media item

>      as per the original endpoints.

Reviewer: Are these hard requirements? If so, it would solve a lot of problems relating to the multipart body (namely issues figuring out which parts are actually useful).

The requirements being:

1. MUST have exactly 2 parts.
2. First part MUST be the JSON object.
3. Second part MUST be the media item.

If these are requirements, the bit about `name` in https://github.com/matrix-org/matrix-spec-proposals/pull/3916/files#r1320678847 is irrelevant.

Author: I rather envisaged that they would be hard requirements, yes.

     No properties are yet specified for the JSON object to be returned.

     An example response:

     ```
     Content-Type: multipart/mixed; boundary=gc0p4Jq0M2Yt08jU534c0p

     --gc0p4Jq0M2Yt08jU534c0p
     Content-Type: application/json

     {}

     --gc0p4Jq0M2Yt08jU534c0p
     Content-Type: text/plain

     This media is plain text. Maybe somebody used it as a paste bin.

     --gc0p4Jq0M2Yt08jU534c0p

>      ```

Reviewer: ftr, the diff highlighter gives up after this code block. Presumably it doesn't like that it's part of a list.

     
5. Backwards compatibility mechanisms

   a. Backwards compatibility with older servers: if a client or requesting
   server receives a 404 error with a non-JSON response, or a 400 or 404 error with
   `{"errcode": "M_UNRECOGNIZED"}`, in response to a request to one of the new
   endpoints, they may retry the request using the original endpoint.
   
   b. Backwards compatibility with older clients and federating servers:
   servers may for a short time choose to allow unauthenticated access via the

>    deprecated endpoints.

Reviewer: The MSC doesn't appear to be requiring authentication on the deprecated endpoints anyways?

Author: Yes. I think the point is that we'll soon have to turn off the deprecated endpoints altogether, otherwise there is no benefit from the new ones. Agreed that this is not entirely clear in the MSC wording.

Reviewer: Deprecation automatically qualifies the endpoints for removal in a later MSC: https://spec.matrix.org/latest/#deprecation-policy

Removal fairly quickly would help encourage servers (and clients) to switch over (hopefully).

Reviewer: I think I've clarified this.

referenced the media, and the time that the link was generated, in the URL.
Such mechanisms would allow the bridge to impose controls such as:

* Limiting the time a media link is valid for. Doing so would help prevent
  visibility to users who weren't participating in the chat.

* Rate-limiting the amount of media being shared in a particular room (in other
  words, avoiding the use of Matrix as a Warez distribution system).

#### Icons for "social login" flows

When a server supports multiple login providers, it provides the client with
icons for the login providers as `mxc:` media URIs. These must be accessible
without authentication (because the client has no access token at the time the
icons are displayed).

This remains a somewhat unsolved problem. Possibly the clients can continue
to call the legacy `/_matrix/media/v3/download` URI for now: ultimately this
problem will be solved by the transition to OIDC. Alternatively, we may need
to provide an alternative `/_matrix/client/v3/login/sso/icon/{idpId}` API
specifically for access to these icon.

(This was previously discussed in
[MSC2858](https://github.com/matrix-org/matrix-spec-proposals/pull/2858#discussion_r543513811).)

> ## Potential issues

Reviewer: How are we envisioning something like [MMR](https://github.com/turt2live/matrix-media-repo) working in this MSC (and MSC3911)? MMR can't use its own domain because it wouldn't have any users in the room (or at all), and it might not be able to reasonably get access/generate a signing key for its underlying domains. Specifically, it doesn't feel great writing "please provide a path to your signing key" into documentation, as the typical environment MMR is deployed in means increasing surface area where signing keys live.

Generating a signing key for an underlying server isn't possible because servers don't support a way to append additional active signing keys to `/_matrix/key/v2/server`, and they probably wouldn't want to anyways because an attacker could add their key to that config[^1]. 

Aside from the surface area concern, signing keys come in a wide variety of file formats. MMR would then need to publish a compatibility matrix that it likely can't adhere to because a server implementation decided to change its key format and break MMR.

There's probably some options where a server can (somehow) advertise that requests from a given domain are expected/support, but that becomes complicated very quickly. It also feels inappropriate for the spec to have first-party support for forward proxies (as this is effectively what the feature would look like).

I do think it's important that the media endpoints be authenticated, and I recognize that MMR may need significant refactoring to support such authentication, but I feel it's also important to consider how deployments like MMR would be expected to work under this MSC. Maybe I'm wrong about the surface area concern, for example - this would make things easier.

For the time being, I'll add a config option to specify the signing key path (and maybe what format it's in), and MMR can just use that. This at least unblocks development for implementation, at risk of security concerns.

[^1]: There are other problems if an attacker can gain access to a homeserver config, but in general a *supported* way to append additional signing keys is very not-great.

Author: You're not entirely explicit about the concern, but if I understand correctly, the problem is that it is difficult for MMR to sign federation requests because it doesn't have access to the signing key belonging to the server it is acting on behalf of?

In which case, the short answer is: I don't know. But let's reason thus:

 * We think it's important to restrict media access to users who have a legitimate reason to see it - ie, that they were in the room at the time of the event.
 * In practice, that means if I have a piece of media, I should only serve it to servers that can see the event.
 * So I need to know which server is asking me for a piece of media.
 * So if a homeserver chooses to delegate media management to a separate service, then it needs a way to authorise that service to make media download requests on its behalf.

So, let's talk about ways to make that work securely.

---

> Generating a signing key for an underlying server isn't possible because servers don't support a way to append additional active signing keys to `/_matrix/key/v2/server`, and they probably wouldn't want to anyways because an attacker could add their key to that config.

I think there are ways to make this work without changes to the spec. For example, if MMR had its own signing key, and exposed a `GET /public_signing_key` endpoint, then I could configure my homeserver to make a request to whichever instance of MMR I am using and include the returned keys in the response to `/_matrix/key/v2/server`.

Indeed, I could have a third service which runs in front of my HS implementation, and proxies  `/_matrix/key/v2/server` requests to both MMR and the HS. Conceptually this makes some sort of sense because we are asking "which signing keys are used by services on this domain" rather than specifically "which signing keys are used by this homeserver".

```

                                                       +------------+
https://domain1/.../v2/server        +---------------> | domain1 HS |
             +-------------------+   |                 +------------+
-----------> | signing key proxy | ->+
             +-------------------+   |
                                     +------> +------------+
                                              | shared MMR |
https://domain2/.../v2/server        +------> +------------+
             +-------------------+   |
-----------> | signing key proxy | ->+
             +-------------------+   |                 +------------+
                                     +---------------> | domain2 HS |
                                                       +------------+
```

The main problem with this is that, just because I want to authorise MMR to make media requests on behalf of my domain, doesn't necessarily mean I want it to make room join requests, sign events, and everything else on my behalf. So perhaps a job for a future MSC would be to replace `/_matrix/key/v2/server` with a new API which allows more granular controls over what a key can be used for.

---

Alternatively: I guess I could host some service (be it part of the HS or otherwise) which effectively proxies and signs requests from MMR, and MMR would have to make federation requests through that service. I'd still need some way to authenticate requests from MMR. It feels a bit clunky.

Reviewer: Thanks for working through my badly worded comment 😅 

Exposing a `/signing_key` endpoint is something I hadn't considered before, but does make sense. MSC3911 already requires extensive homeserver cooperation with MMR, so another endpoint wouldn't be the end of the world. It also avoids concerns about running middleware proxies all over the place.

Will give it a go and see how it feels.

Reviewer: Just to follow up here (and resolve the thread), in the process of implementing the above in MMR it turns out that Synapse supports multiple signing keys already, removing the need for a dedicated API. MSC3911 might also not need *extensive* cooperation as nearly all checks/calls can happen over existing Client-Server API endpoints. Dedicated API might be more performant, but ultimately not required.

Synapse's multiple key support looks like an ancient feature from the early days of development, but still works trivially well. There's no obvious way to generate such a signing key "properly", but if the signing key file has multiple keys in it then Synapse will use them all to service `/_matrix/key/v2/server`. Only the first key is used by Synapse itself to sign events and outbound requests.

This setup means that MMR can generate a signing key that it can use, ask the user to combine it with their existing Synapse signing key, and restart their homeserver. Synapse will automatically pick up the second signing key, and not use it for it's own activities (outside of `/_matrix/key/v2/server). Such characteristics give the admin the option to restore their signing key from a backup if they ever need to revoke MMR's signing key, without having to specify it as an old verify key (because it was never used in a persistent manner). An admin could also instruct MMR to use the server's existing signing key, but then they lose the ability to revoke MMR's signing key easily. 

Overall, this sort of setup feels fine enough from a security and install experience perspective imo, and very much out of scope for this MSC. It's arguable that the dual signing key support is an accidental/unsupported feature of Synapse (particularly considering Synapse doesn't generate such keys in normal operation), however this would hardly be the first time MMR relies on Synapse's internals to get something done. Other homeservers, namely Dendrite, do not support this sort of feature at all either, so MMR is already going to have to consider what a [partially?] proxied `/_matrix/key/v2/server` endpoint would look like - if Synapse's internals change drastically enough, the proxy approach would also work for Synapse users.

Resolving as this is out of scope for the MSC, and was from the begining (not sure why I posted it here in the first place...).

> ## Potential issues

Reviewer: This MSC does not allow for use of the `Range` header over federation, as it's unclear which "part" the caller could use as a range. I don't believe any servers use range requests, but many clients do on the existing combined endpoint.

Reviewer: it's not entirely clear to me that a server over federation would want to use `Range` requests, fwiw. Similarly, it's not obvious to me that a server would use `/thumbnail` over federation, but the argument would be that if `/thumbnail` is exposed as a "value add" feature then `Range` might also be good for a similar quality of life enhancement.

Author: > it's not entirely clear to me that a server over federation would want to use Range requests, fwiw. 

well, it might be useful to resume broken downloads?

> Similarly, it's not obvious to me that a server would use /thumbnail over federation,

Hrrrm, a very interesting point. Synapse doesn't do so (cf https://github.com/matrix-org/synapse/blob/v1.92.3/synapse/rest/media/thumbnail_resource.py#L275-L277 which suggests that 2014-era Mjark thought that maybe we *should* use `/thumbnail`), but... if there's no usecase for it, maybe we should get rid of `GET /_matrix/federation/v1/media/thumbnail` while we're changing things here.

Reviewer: MMR also doesn't use the thumbnail endpoint over federation. It's just too uncomfortable to use: the origin server can create arbitrary images and call them thumbnails. 

This is also possible with the download endpoint, but tracking a download hash is easier than a thumbnail hash (when we eventually do that in another MSC)

Reviewer: I'm fairly sure the HTTP spec doesn't actually describe how to deal with this situation, and it turns out that Synapse, MMR, etc all discard received bytes and fail the associated client requests upon federation download failure - meaning there's no resumed download - so I'm leaving `Range` request handling as a "specification interpretation detail" in this MSC. 

Resolving as this will be documented in the MSC itself.

Reviewer: re: thumbnail. I'm just going to remove the federation endpoint, but will do so in a different commit.

> ## Potential issues

Reviewer: This MSC is incompatible with the principles of https://github.com/matrix-org/matrix-spec-proposals/pull/3860 (being able to host media in a CDN) - the federation endpoints are unable to be redirected due to the multipart form body (I believe).

A workaround might be to have the second part contain a `Location` header for redirection.

Author: I've got to say that the principles of this MSC - and more particularly MSC3911 - are somewhat at odds with hosting media in a CDN anyway. I think we need to consider how the two requirements interact.

> A workaround might be to have the second part contain a Location header for redirection.

yeah. Or spec a property for the json part. 

Or, possibly, another vote for just using a separate endpoint.

Author: > This MSC is incompatible with the principles of https://github.com/matrix-org/matrix-spec-proposals/pull/3860 (being able to host media in a CDN) - the federation endpoints are unable to be redirected due to the multipart form body (I believe).

It's not so much that the federation endpoint couldn't be redirected, but that a CDN might not be able serve the response in the right format (I suspect multipart/mixed responses aren't very cacheable), or honour the authentication requirements.

Reviewer: The principles of requiring authentication I don't think are totally at odds with CDNs, though they don't traditionally work together very well. At least in AWS S3, it's possible to generate a time-limited URL that's still part of the wider CDN, allowing an application to authenticate the request before generating such a URL it can redirect to.

Whether this actually saves any time/money in practice is unknown. I suspect it does save *some* money, but not as much as today's unauthenticated endpoints do.

I mention MSC3860 because for better or worse it's merged to the spec today, which unfortunately means this MSC needs to consider it as well. Ideally the order of operations would have been the other way around :(

Reviewer: ftr, I was chatting with some CDN operators at the Matrix fringe event during FOSDEM 2024, and together we came up with https://github.com/matrix-org/matrix-spec-proposals/pull/4097 as a possible fix for the problem.

Reviewer: CDNs can also rewrite the response body in most cases, so the JSON could be provided as a query string parameter and still cache the underlying media type. 

I've documented this in the MSC.

Reviewer 3: > Or, possibly, another vote for just using a separate endpoint.

+1 to a separate endpoint for media metadata over federation because:
 - I'm unsure how widely supported multipart/mixed is in practice. I had never heard of it until now.
 - Making multiple http requests is cheap if you're using HTTP/2 or 3, the only downside would be an extra round trip which could be done in parallel if you wanted.
 - Keeping metadata separate from content allows HEAD style requests which could potentially be useful. E.g it could say which rooms the media belongs in, or which users have access, without having to then download the content.

* Setting the `Authorization` header is going to be annoying for web clients. Service workers
  might be needed.

* Users will be unable to copy links to media from web clients to share out of
  band. This is considered a feature, not a bug.

> ## Alternatives

Reviewer: Needs comparison to https://github.com/matrix-org/matrix-spec-proposals/pull/2461 and https://github.com/matrix-org/matrix-spec-proposals/pull/701

> * Allow clients to upload media which does not require authentication (for

Reviewer 4: I think allowing some form of public upload (via query string or different path) should be required, this is not only useful for IRC/XMPP bridges but also for public rooms (like MatrixHQ). When media is available publically makes it possible to use CDN/edge caching much more effectively, which is critical if you want media to be fast worldwide. This is particularly important in large public rooms. Perhaps this should be optional?

Author: The difficulty here is that, frankly, it still leaves the door open for matrix users to treat their homeserver as a free cloud data store, along with a bunch of the other problems (in particular that it's hard to delete media, in response to redaction/GDPR/everybody leaving the room/etc). Being able to tie pieces of media to specific events is really useful.

I do hear the concern that this could be a significant performance problem in larger rooms though :/

Reviewer: A server might still be able to handle this by checking to see if the media is in a public room (or uploaded with an implementation-specific flag to denote public media). If the media is "public" by the server's standards, it can skip the auth checks and return the media.

This behaviour wouldn't be sanctioned by the MSC or the spec, but it is a possible implementation-level efficiency choice. Requests for media would still need to have authentication metadata, but a server (or CDN) can simply discard it upon receipt. Preferably though, the server would at least check to ensure the credentials are of legal shape/probability.

Reviewer: (resolving - the above feels like it addresses the concern, and followup would likely be better placed on MSC3911 instead)

  rename, and it is expected that both proposals will be mergeed at the same
  time (so a double rename will be confusing and unnecessary). However, if
  MSC3911 is delayed or rejected, we should reconsider this.

* Rather than messing with multipart content, have a separate endpoint for
  servers to get the metadata for a media item. That would mean two requests,
  but might make more sense than both `/thumbnail` and `/download` providing
  the info.

## Unstable prefix

While this proposal is in development, the new endpoints should be named as follows:

* `GET /_matrix/client/unstable/org.matrix.msc3916/media/preview_url`
* `GET /_matrix/client/unstable/org.matrix.msc3916/media/config`
* `GET /_matrix/client/unstable/org.matrix.msc3916/media/download/{serverName}/{mediaId}`
* `GET /_matrix/client/unstable/org.matrix.msc3916/media/download/{serverName}/{mediaId}/{fileName}`
* `GET /_matrix/client/unstable/org.matrix.msc3916/media/thumbnail/{serverName}/{mediaId}`
* `GET /_matrix/federation/unstable/org.matrix.msc3916/media/download/{serverName}/{mediaId}`
* `GET /_matrix/federation/unstable/org.matrix.msc3916/media/thumbnail/{serverName}/{mediaId}`

> 

Reviewer 5: If you need any implementations, I currently have a very similar patch in use on my server. It just adds authentication to the existing endpoints in synapse, but you can implement this MSC by just running a find and replace on the paths/doing some changes in the reverse proxy:

- Synapse: https://github.com/deepbluev7/synapse/commit/475015463126dc6dc488fa838039b7cfeced5435
- Nheko also already sends auth headers when downloading media files, as do a few other clients (but still using the legacy paths).

It's not a full implementation, but someone could take that and make one from it and it demonstrates the feasibility. I guess seeing an implementation in a bridge or webclient would be useful though. It currently looks more like the second alternative mentioned, but I don't think being able to change those paths should be controversial.

Reviewer: Thank you for taking the time to write this up. I've linked some modern implementations in the MSC description, and more are expected over the next few weeks. 

I'm resolving this for now as more complete implementations have been referenced.

   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}/{fileName}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaidfilename) | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}/{fileName}` | -                                                                 |
   | [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)          | `GET /_matrix/client/v1/media/thumbnail/{serverName}/{mediaId}`           | - |

   **Note**: [`POST /_matrix/media/v3/upload`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixmediav3upload)
   is **not** modified by this MSC: it is intended that it be brought into line with the other
   endpoints by a future MSC, such as [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).
   It is subsequently **not** deprecated either.

   **Note**: `/thumbnail` does not have a federation endpoint. It appears as though
   no servers request thumbnails over federation, and so it is not supported here.

>    A later MSC may introduce such an endpoint.

Reviewer 6: Not sure if having a potential regression in this MSC is the best idea, considering that `/thumbnail` could be helpful. For example, if you're scrolling through a timeline of videos and want to see the thumbnails, but not actually play, and hence download, all the content it would be prevent unnecessary downloads to the media repo.

Requiring a new MSC for this would significantly increase the friction of future usage.

Reviewer: That all doesn't happen over federation though - the client-server API endpoint for thumbnails is still retained.

Reviewer 6: Yes, but there is clearly potential value in doing so over federation in the future. I don't mind *that* much if the federation endpoint isn't implemented now, but it would still be nice to have. :slightly_smiling_face:

Reviewer: No servers have used it over federation, so I'm inclined to keep this as a consideration for a future MSC.

Reviewer 6: Ok, you can mark this as resolved then. (this was my final concern with the MSC as well)

Reviewer 7: I think if it is valuable in the future a separate MSC can make that argument to reintroduce it!

Reviewer 8: > No servers have used it over federation, so I'm inclined to keep this as a consideration for a future MSC.

This is untrue, conduit [does this](https://gitlab.com/famedly/conduit/-/blob/7b19618136184382560abd2c13cc5c374b0b1b6a/src/api/client_server/media.rs#L201-213) if the full file isn't already available locally and has been doing so [for three years](https://gitlab.com/famedly/conduit/-/commit/aa5e9e607ecc739d0d991ea7221dadd0125f6d64). How was this missed?

Reviewer 8: (re-posted as a review rather than a single comment)

> No servers have used it over federation, so I'm inclined to keep this as a consideration for a future MSC.
 
This is untrue, conduit [does this](https://gitlab.com/famedly/conduit/-/blob/7b19618136184382560abd2c13cc5c374b0b1b6a/src/api/client_server/media.rs#L201-213) if the full file isn't already available locally and has been doing so [for three years](https://gitlab.com/famedly/conduit/-/commit/aa5e9e607ecc739d0d991ea7221dadd0125f6d64). How was this missed?

Reviewer: Good catch, and thanks for raising! In a brief discussion with both the SCT and Conduit devs, the best course of action for now appears to be reinstating `/thumbnail` to the Federation API. A later MSC may still deprecate/remove it, but for now the SCT thinks it's fine enough to include.

Reviewer: Implementation: https://github.com/t2bot/matrix-media-repo/pull/509/commits/07cdf7beac278938846494b3966471d98f6d77bd

   encouraging updates to those affected projects as needed.

2. Removal of `allow_remote` parameter from `/download`

   The current
   [`/download`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaid)
   and
   [`/thumbnail`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)
   endpoints take an `allow_remote` query parameter, indicating whether the
   server should request remote media from other servers. This is redundant
   with the new endpoints, so will not be supported.

   Servers MUST never return remote media from `GET /_matrix/federation/v1/media/download`;
   indeed, the `serverName` is included in the URI only for consistency with the CS-API.

   `/_matrix/client/v1/media/download` and
   `/_matrix/client/v1/media/thumbnail` return remote media as normal.

3. Authentication on all endpoints

   Currently, the `/download` and `/thumbnail` endpoints have no authentication
   requirements. Under this proposal, the new endpoints will be authenticated
   the same way as other endpoints: they will require an `Authorization` header

>    which must be `Bearer {accessToken}` for `/_matrix/client`, or the signature

Reviewer 6: :nerd_face: Query params have not been removed from the spec for C-S auth, so you should either update this MSC to reflect that, or just remove this explicit explanation of auth.

Reviewer: I think we're explicitly saying that the query string `access_token` approach is explicitly not supported here - will clarify.

* We could simply require that `Authorization` headers be given when calling
  the existing endpoints. However, doing so would make it harder to evaluate
  the proportion of clients which have been updated, and it is a good
  opportunity to bring these endpoints into line with the rest of the
  client-server and federation APIs.

* There's no real need to rename `GET /_matrix/media/v3/preview_url` and `GET
  /_matrix/media/v3/config` at present, and we could just leave them in
  place. However, changing them at the same time makes the API more consistent.

  Conversely, we should make sure to rename `POST
  /_matrix/media/v3/upload`. The reason to delay doing so is because MSC3911
  will make more substantial changes to this endpoint, requiring another
  rename, and it is expected that both proposals will be mergeed near to the same
  time as each other (so a double rename will be confusing and unnecessary). However,
  if MSC3911 is delayed or rejected, we should reconsider this.

* Rather than messing with multipart content, have a separate endpoint for
  servers to get the metadata for a media item. That would mean two requests,

>   but might make more sense than `/download` providing the info directly.

Reviewer: Early evidence shows it's harder for some languages to support multipart bodies as described by the MSC, so it may be in our best interest to split the endpoint up after all.

Reviewer: This ended up being mostly a non-issue. The concern was that (Twisted/)Python might have a hard time with this sort of thing, but the implementation turned out fairly sane and not wildly complicated: https://github.com/element-hq/synapse/pull/17172

Reviewer: @kegsay [says](https://github.com/matrix-org/matrix-spec-proposals/pull/3916/files#r1600677745):

> > Or, possibly, another vote for just using a separate endpoint.
> 
> +1 to a separate endpoint for media metadata over federation because:
> 
> * I'm unsure how widely supported multipart/mixed is in practice. I had never heard of it until now.
> * Making multiple http requests is cheap if you're using HTTP/2 or 3, the only downside would be an extra round trip which could be done in parallel if you wanted.
> * Keeping metadata separate from content allows HEAD style requests which could potentially be useful. E.g it could say which rooms the media belongs in, or which users have access, without having to then download the content.
> * having a separate endpoint enables bulk lookups which could be useful for say fetching all media metadata in a given room, for populating an Attachments / Uploaded Files list.

Reviewer: cross-ref concern on this subject: https://github.com/matrix-org/matrix-spec-proposals/pull/3916#issuecomment-2112945365

Reviewer 9: Having thought about it more: sending the whole multipart blob to the CDN doesn't make any sense because clients downloading the media don't want that data. IMO the spec should keep allowing the use of "dumb" CDNs that just serve bytes with some generic presigned URLs. The multipart content type would effectively require CDNs to be smart

Sending the data in a header seems like an easy solution for all the problems. The only potential new issue I can think of is size limits. The HTTP spec doesn't seem to define a limit, but searching indicates that the limit is at least 4kb across all common webservers, which sounds high enough for a bit of permission metadata.

Reviewer 10: I'm not hugely in favour of shoving metadata into headers, as care then needs to be taken to ensure such headers pass through any proxies etc.

Some thoughts on other points:

> I'm unsure how widely supported multipart/mixed is in practice. I had never heard of it until now.

Multipart is fairly widespread, `form-data` and email being the prime examples. The format is also fairly easy, so I wouldn't be particularly concerned about language support.

> IMO the spec should keep allowing the use of "dumb" CDNs that just serve bytes with some generic presigned URLs. The multipart content type would effectively require CDNs to be smart

Using headers on a redirect response is a mild nuisance, as you can't then use the built-in redirect handling of most HTTP client libraries. This removes one of the nice properties of just allow redirects (it will work most of the time automatically).

An alternative would be for the federation response to allow specifying a URL to fetch the data from in the returned metadata. This would essentially be equivalent to a redirect, but without having to shove things in headers.

Reviewer 9: Custom redirects would be sufficient for me, although they wouldn't help with the two other points (`Range` and `HEAD` requests). I'd prefer headers as I feel they're better overall even if you have to handle the first redirect manually, but I don't feel too strongly about it.

<details>
<summary>For reference, the places where I use redirects are:</summary>

* Beeper's media repo: normal media repo backed by a relatively dumb CDN.
* Discord direct media: special media repo built into the bridge that only redirects to discord. Can do custom redirect responses, but can't place any new requirements on the CDN for obvious reasons.
* Giphy direct media: currently implemented in [7 lines of Caddyfile](https://mau.dev/maunium/caddy/-/blob/master/vhosts/giphy.mau.dev#L4-10). Custom redirect responses would be more complicated, while adding a header with a static value would be trivial.

</details>

Reviewer: > I'm not hugely in favour of shoving metadata into headers, as care then needs to be taken to ensure such headers pass through any proxies etc.

The default behaviour of proxies is to pass the header along in a response. Request headers are less likely to work, but we aren't using those here.

Reviewer 10: > The default behaviour of proxies is to pass the header along in a response. Request headers are less likely to work, but we aren't using those here.

For standard HTTP proxies sure, but we have other things like SBGs and the like. And it's not just about passing through, its also about ensuring that they don't mangled or e.g. size limited.

Reviewer 3: > Multipart is fairly widespread, form-data and email being the prime examples. The format is also fairly easy, so I wouldn't be particularly concerned about language support.

My concern around `multipart/mixed` is that it generally isn't as widely supported as other `multipart` types like form data, not that multipart in general is not widespread. Some data points:
- Go didn't have multipart/mixed support until 2018 https://github.com/golang/go/issues/23959
- I can't seem to find any explicit multipart/mixed support in Rust's Axum or Hyper, but I'm sure I'm just looking in the wrong place, though I'm [clearly not alone](https://users.rust-lang.org/t/http-batch-request-response-multipart-mixed-content-type/31834).

I'd rather we split metadata + raw data for the reasons I outlined earlier, hence leaning more towards split endpoints rather than using HTTP headers.

Reviewer 10: > My concern around multipart/mixed is that it generally isn't as widely supported as other multipart types like form data, not that multipart in general is not widespread.

I mean, they're all the same format it just changes the semantic meaning of how the different section related. [Axum](https://docs.rs/axum/latest/axum/extract/struct.Multipart.html) has inbuilt support (note that it doesn't seem to actually check if the header is form-data vs other multipart, despite only talking about `form-data`), it just uses [multer](https://docs.rs/multer/latest/multer/) under the hood.

> I'd rather we split metadata + raw data for the reasons I outlined earlier, hence leaning more towards split endpoints rather than using HTTP headers.

I don't know, something about having to do multiple HTTP requests to get the data and metadata just feels as bit needless to me (especially since we don't wildly use HTTP/2 for federation)  :shrug:

Reviewer 7: > Some good news is there is some precedent for including JSON in a header via [`NEL`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/NEL), so we can likely do the same without feeling bad about it.

Would it make sense to have multiple headers for different bits instead of shoving JSON into a header? E.g. for MSC3911 it could be:

`Matrix-Restrictions: event_id=$Rqnc-F-dvnEYJTyHq_iKxU2bZ1CI92-kuZq3a5lr5Zg`

OR

`Matrix-Restrictions: profile_user_id=@clokep:matrix.org`

If there isn't a *huge* potential for additional fields then multipart/mixed with a JSON object seems like major overkill.

Reviewer 9: Multiple headers would make adding any new metadata much harder: servers might fetch a piece of media before implementing new metadata and therefore never store it. If it's all in one json object, servers can store the whole object without recognizing what's inside and then start applying the metadata after it's implemented.

Reviewer: It feels like there's equal upsides and downsides to each approach, which I've tried to illustrate below with heavy bias. The TLDR is I don't think the MSC *needs* to change, and any of the options are fine enough for Matrix. I'll bring this block of text into the MSC after some discussion here on the thread.

Thoughts welcome.

### Option 1: Use `multipart/mixed`

Pros:
* Single neatly packaged endpoint.
* For servers not using http/2, reduced connection overhead.
* Relatively easy to parse.

Cons:
* CDN-backed servers may have a hard time supporting it. The homeserver could append some information to the redirect URL for an edge worker at the CDN to parse and manipulate the response format. The information could be signed as well, to help ensure the metadata isn't falsified by an intermediary client.
* Federation requests would not be able to use `Range` headers. No servers use `Range` over federation today.
* Using an upstream CDN, like how the mautrix-discord bridge works, becomes impossible without proxying somewhere. This is a relatively uncommon practice, and in my opinion doesn't feel like something we should optimize for. An option may be to use an intermediary CDN which proxies the upstream CDN in addition to the behaviour described above for generic CDNs.
* Some languages have a harder time parsing the response. It's relatively simple to write a tokenizer for, though.

### Option 2: Split endpoints

Pros:
* `Range` requests become supported. Though, no servers will use them.
* A variety of CDN use cases become more possible.
* Language support is already present.

Cons:
* Servers not using http/2 will see increased connection counts. The default in Matrix is http/1.1.
* Generally increased complexity and opportunity for failure.
* Increased maintenance surface for Matrix and its implementations.

Notes:
* An unlisted 'pro' is being able to retrieve metadata independently of the media, allowing a server to check for access without spending a bunch of time on bandwidth. This isn't an optimization the server can perform: with MSC3911 in effect, the requesting server would receive an *error* if any of its users can't see the media being downloaded. There is potentially some wasted bandwidth/time spent if the local user requesting the media doesn't have access, though implementations could interrupt the response stream after reading (and validating) the metadata chunk too. Servers may find it useful to at least continue caching the media for whichever user *does* have access, too.
* Another unlisted 'pro' is an attachments list functionality. Under MSC3911, this would be constructed using events rather than media requests.
* An unlisted 'con' is developers may be confused when they try to `GET /download` and receive an access error. This wouldn't come up in local development, but would in a production environment. Documentation in the spec to say "hey, download the access metadata [here]" could help, though may still be overlooked. This is a relatively insiginificant problem, but is opportunity for improved developer experience in Matrix.

### Option 3: Move metadata to custom header(s)

Pros:
* We can safely shove JSON into a header without feeling bad.
* Supports `Range` requests.
* Supports CDNs with edge worker support.

Cons:
* Bridges redirecting to upstream CDNs still don't work without proxying the media somewhere.
* Some forward and reverse proxies will strip the headers, breaking functionality. Application-specific proxies may update, though is increasingly specialized and harder to maintain.

Reviewer 9: I think redirects and dumb CDNs do need to be supported, but I don't really care how exactly it's implemented. If we want to stick with multipart as the default method, then supporting redirects using a new field would be fine.

If the MSC is accepted and implemented without redirect support, then adding it later will require another migration system like `allow_redirect` last time, but this time it's completely pointless and easily avoidable.

> Bridges redirecting to upstream CDNs still don't work without proxying the media somewhere.

Upstream CDNs would work fine with redirects and headers, unless the headers are required to be on the redirect target rather than the first response

Author: >  An unlisted 'con' is developers may be confused when they try to `GET /download` and receive an access error. This wouldn't come up in local development, but would in a production environment.

I'm struggling to follow this through. Presumably `/download` here is the endpoint that returns the media, not the metadata? Why would you get an access error? Why would you not get one in development? (Also, why are these pros and cons "unlisted"?)

Author: > If the MSC is accepted and implemented without redirect support, then adding it later will require another migration system like allow_redirect last time, but this time it's completely pointless and easily avoidable.

I'm not quite following this either. Remind me what `allow_redirect` refers to? What would "supporting redirects using a new field" look like?

> Upstream CDNs would work fine with redirects and headers, unless the headers are required to be on the redirect target rather than the first response

Unfortunately I think the expectation *would* be that the headers are on the final response.

Reviewer 9: > I'm not quite following this either. Remind me what `allow_redirect` refers to?

Redirects were originally added in Matrix 1.7 (MSC3860) and there's an `allow_redirect` query param to opt into redirects, because not everything followed redirects before the spec change.

If redirect support is included in this MSC, no opt-in is necessary as all servers will implement it correctly from the beginning. If it's not included, then a future MSC to re-add redirects would again have to add an opt in mechanism.

> What would "supporting redirects using a new field" look like?

Returning a download link instead of the file in the multipart response. Probably makes sense to have a new multipart field for that, but just putting it in the json metadata could work too. The reason *not* to put redirects in metadata is that the links are usually temporary, not something that should be persisted

Does multipart/mixed allow other headers than `Content-Type` in the parts? If yes, then we could just have a `Location` header instead of the content part:

```mime
Content-Type: multipart/mixed; boundary=gc0p4Jq0M2Yt08jU534c0p

--gc0p4Jq0M2Yt08jU534c0p
Content-Type: application/json

{}

--gc0p4Jq0M2Yt08jU534c0p
Location: https://cdn.example.com/file?signature=foo

--gc0p4Jq0M2Yt08jU534c0p
```

If not, then maybe putting it in the json object is best, implementations just need to be careful not to store that field.

Reviewer: I'm not sure I understand where the `allow_redirect` subthread is coming from. The MSC already strips `allow_redirect` to make it a first-class feature of the endpoint, regardless of this thread's outcome regarding multipart, split endpoints, or headers. If that language isn't clear, please raise a dedicated thread for it.

Reviewer: Regarding redirect behaviour: I am fairly strongly opposed to redesigning HTTP redirects in Matrix, and favour modifying our use case for supporting 307/308 redirects. Custom redirects add high maintenance burden to the spec and its implementations. This does mean that headers (if we were to use that option) will need to be on the redirect target rather than the intermediary.

Reviewer: > > An unlisted 'con' is developers may be confused when they try to `GET /download` and receive an access error. This wouldn't come up in local development, but would in a production environment.
> 
> I'm struggling to follow this through. Presumably `/download` here is the endpoint that returns the media, not the metadata? Why would you get an access error? Why would you not get one in development? (Also, why are these pros and cons "unlisted"?)

@richvdh - `/download` here is just media rather than metadata, yes. With MSC3911, it's possible the server would not have visibility on the event the media was shared in, thus resulting in error. This feels like an edge case which might not be caught in development, but would (eventually) happen in production.

Those pros and cons are "unlisted" because they don't have terribly strong arguments. It could be considered a non-issue that developers run into edge cases because they didn't read the spec, for example.

Reviewer 10: > This does mean that headers (if we were to use that option) will need to be on the redirect target rather than the intermediary.

Which basically means that we can't use standard CDNs right? Isn't that the crux of the issue here

Reviewer: It may depend on the definition of "standard", I think. Cloudflare, Fastly, and all the other major vendors support an "edge worker" (naming varies) which can interpret a request, do some logic, then serve a different response format. 

The flow for such a worker would be:

1. Homeserver receives federation request to download media
2. Homeserver identifies that a CDN actually has the media instead of it
3. Homeserver generates a URL containing:
   * CDN location and appropriate pathing
   * Metadata restrictions (likely stored in a query string)
   * Information about which servers have access. This could be narrowly scoped to the single requesting server (ie: time-limited URL) or broadly scoped to heavily cache the response and reduce processing on the CDN edge (ie: `?allow=matrix.org&allow=example.org`, though probably encoded/encrypted for privacy reasons)
   * Signature to cover the metadata restrictions, access information, and pathing.
4. Homeserver responds with a 307 or 308 redirect to the generated URL
5. Client (remote server) is redirected, including the `X-Matrix` auth header(s).
6. CDN verifies the `X-Matrix` auth, compares the server against the access controls included in the signed URL, and responds with errors as appropriate.
    * To avoid the edge also doing key lookups, the homeserver could include (and sign) the `X-Matrix` headers it saw and validated so the CDN can simply do a string comparison rather than needing to go hit `/_matrix/key/v2/server` everywhere.
7. If all the access controls look good, including the auth headers, the edge worker then retrieves the object from the underlying store (typically S3 or a file server of some kind), prepends metadata information (as included in the signed URL), and starts streaming a response.

A lot of this is implementation-specific, but that's the sort of thing https://github.com/matrix-org/matrix-spec-proposals/pull/4097 aims to consider. (note: I'm considering closing 4097 as Not Our Problem given the high implementation specificity here).

None of this works if the homeserver is trying to do a simple redirect to Discord's CDN, though I don't believe this should be a requirement. I'm struggling to articulate why I think this, but it feels like improper form to support such functionality when we're locking down on it in Matrix via this MSC.

Reviewer 9: Even with smart CDNs, that seems like a lot of extra work. I'd much rather implement a bit of extra complexity on the requesting side in order to have an API that can redirect to anything. A custom redirect in the multipart response would be relatively trivial, split endpoints shouldn't be too hard either

> it feels like improper form to support such functionality when we're locking down on it in Matrix via this MSC.

From my point of view, media permissions are being locked down because media repos of normal servers need it for trust & safety reasons, but I don't think there's anything wrong with allowing other kinds of media repos (like 3rd party CDN redirects)

In the end, restricting redirects won't prevent any use cases, it would really just require some kinds of media repos to waste bandwidth on proxying everything.

Reviewer: I'm not sure it's correct to move the complexity into Matrix for what appears to be an edge case. 

I'm also hoping to get this thread resolved though so the MSC can get put through the process quickly. Are folks more in favour of split endpoints or multipart? (The other options already appear discounted, but happy to hear new points on them too)

Reviewer 10: @turt2live it might help if you expanded on the complexity custom redirects have? I know there is some complexity, but it may very well be less than the other options?

Personally I'm fairly in favour of multipart, the pros for doing split out endpoints seem mainly theoretical.

Reviewer: > @turt2live it might help if you expanded on the complexity custom redirects have? I know there is some complexity, but it may very well be less than the other options?

It's added spec which requires fairly strict behaviour, and added code which likely needs tests and maintenance. HTTP meanwhile has already specified a redirection mechanism, and we should delegate that responsibility where we can in my opinion. It may not be the perfect mechanism for us, but it avoids having to re-specify how a redirect works in applications. Standard HTTP redirects also generally have library support, making the redirect invisible to the application/developer.

A more vague concern is the custom redirect handling code would essentially be the same as splitting the endpoints. If the first call results in a redirect, make a second call (potentially with a different request shape). The only difference code-wise would be that with split endpoints the first call needs to be a 200 OK whereas the custom redirect is 200 OK *and* redirect headers.

Reviewer 10: > > @turt2live it might help if you expanded on the complexity custom redirects have? I know there is some complexity, but it may very well be less than the other options?
> 
> It's added spec which requires fairly strict behaviour, and added code which likely needs tests and maintenance. HTTP meanwhile has already specified a redirection mechanism, and we should delegate that responsibility where we can in my opinion. It may not be the perfect mechanism for us, but it avoids having to re-specify how a redirect works in applications. Standard HTTP redirects also generally have library support, making the redirect invisible to the application/developer.

That's fair, though I think the custom redirect spec and code is fairly low cost (though obviously not free). It's basically: "if you the metadata includes `media_url`, then do a standard GET to that to get the media", I don't think its anymore complex than that?

I think the trade offs in my mind then are:

| Approach | Pros | Cons |
| - | - | - |
| Standard redirect with multipart | - Easy redirection behaviour<br/>- Only uses one (or two) HTTP requests | - No dumb CDN support |
| Split out endpoints | - Easy redirection behaviour<br/>- Dumb CDN support | - Requires two (or three) HTTP requests |
| Multipart and custom redirect | - Only uses one (or two) HTTP requests<br/>- Dumb CDN support |  - Redirection must happen manually |

So the pertinent questions are:
1. Do we want to support dumb CDNs? I would say yes (feeling: +0.7)
2. Do we want to reduce number of HTTP requests per fetch? I would say yes (feeling: +0.2)
3. How much work do we think adding custom redirection is? I would say low.

Though I'm mindful that I'm framing the questions to prioritise what I'm currently thinking.

---

On the CDN side I'm making some assumptions:
- That we don't care about latency, as much as offloading the bandwidth cost. 
- I think a common use case would be to also use something like S3 (which supports time-limited URLs), and the like.
- Having edge workers looks convoluted, and not necessarily easy to test. Having lots of deployments re-implement that edge worker logic, so as to avoid the fewer servers and spec from implementing the custom redirect logic, seems slightly perverse.

Reviewer: From brief discussions with the SCT, and the latest comment, it sounds like the vote is largely towards multipart with custom redirects - that will be enshrined shortly into the MSC.

Concerns with the selected approach should be raised as mscbot concerns, please.

Reviewer: Implementation: https://github.com/t2bot/matrix-media-repo/pull/509/commits/2dcd96924ef48cbce51fc80c9404b354f094d79b

   | [`GET /_matrix/media/v3/preview_url`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3preview_url) | `GET /_matrix/client/v1/media/preview_url` | - |
   | [`GET /_matrix/media/v3/config`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3config) | `GET /_matrix/client/v1/media/config` | - |
   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaid)            | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}`            | `GET /_matrix/federation/v1/media/download/{serverName}/{mediaId}`  |
   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}/{fileName}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaidfilename) | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}/{fileName}` | -                                                                 |

>    | [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)          | `GET /_matrix/client/v1/media/thumbnail/{serverName}/{mediaId}`           | - |

Reviewer 7: This table is missing the [`media/v1/create` endpoint](https://spec.matrix.org/v1.10/client-server-api/#post_matrixmediav1create).

```suggestion
   | [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)          | `GET /_matrix/client/v1/media/thumbnail/{serverName}/{mediaId}`           | - |
   | [`POST /_matrix/media/v1/create`](https://spec.matrix.org/v1.10/client-server-api/#post_matrixmediav1create)          | `POST /_matrix/client/v1/media/create`           | - |
```

Reviewer 7: Or maybe this is not modified like `/upload`? But I don't see it mentioned at all.

Reviewer 6: It already uses authentication, so I don't see why it would be modified.

Reviewer: it should be moved into the client namespace, like the URL preview endpoint.

I may apply this suggestion offline because reviewing table diffs is awful on github 😭

Reviewer 7: > it should be moved into the client namespace, like the URL preview endpoint.

Yes, this was my thinking -- just for consistency not because there's changes happening.

> I may apply this suggestion offline because reviewing table diffs is awful on github 😭

I make no guarantee the table is correct. 😄

Author: /create postdates the creation of this MSC (yes it's been that long). Certainly we should update it to match the other endpoints.

   **Note**: [`POST /_matrix/media/v3/upload`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixmediav3upload)

>    is **not** modified by this MSC: it is intended that it be brought into line with the other
>    endpoints by a future MSC, such as [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).
>    It is subsequently **not** deprecated either.

Author: "subsequently"? As in, MSC3911 doesn't deprecate it either? That's not what MSC3911 says.

Possibly you mean "consequently", but I suggest:

```suggestion
   is **not** modified or deprecated by this MSC: it is intended that it be brought into line with the other
   endpoints by a future MSC, such as [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).
```

Reviewer: applied manually

   which support the new `download` and `thumbnail` endpoints SHOULD cease to serve
   newly uploaded media from the unauthenticated versions. This includes media
   uploaded by local users and requests for not-yet-cached remote media. This is
   done with a 404 `M_NOT_FOUND` error, as though the media doesn't exist. Servers
   SHOULD consider their local ecosystem impact before freezing the endpoints. For
   example, ensuring that common bridges and clients will continue to work, and
   encouraging updates to those affected projects as needed.

2. Removal of `allow_remote` parameter from `/download`

   The current
   [`/download`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaid)
   and
   [`/thumbnail`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)
   endpoints take an `allow_remote` query parameter, indicating whether the
   server should request remote media from other servers. This is redundant
   with the new endpoints, so will not be supported.

   Servers MUST never return remote media from `GET /_matrix/federation/v1/media/download`;

>    indeed, the `serverName` is included in the URI only for consistency with the CS-API.

Reviewer 7: This seems like a potential footgun. The only people who may care that the URLs differ is homeserver implementors. I would recommend *not* including the serverName for the federation endpoint -- having it adds additional (useless?) validation.

I could see it being valuable for something like vhosting, but I think you have that info in the hostname header anyway.

Reviewer 9: > I could see it being valuable for something like vhosting, but I think you have that info in the hostname header anyway.

Host headers can be inaccurate due to delegation and proxying, but the `X-Matrix` authorization header (which will now be required) includes a `destination` field as of MSC3383

Reviewer: `destination` is still optional, and we use REST principles in Matrix. I'm not sure this is the MSC where we want to really challenge that 😅

Author: I don't think I feel strongly about this, though reducing the amount of duplication (and hence potential foot-shooting-incidents like https://github.com/matrix-org/synapse/issues/3796) is attractive.

We could *require* `destination` for this endpoint, I suppose.

> we use REST principles in Matrix. I'm not sure this is the MSC where we want to really challenge that 😅

I don't really follow that this would violate REST principles. Do you mean that it's surprising for `GET /_matrix/federation/v1/media/download` to return different results depending on the server? We have that situation for [`GET /_matrix/key/v2/server`](https://spec.matrix.org/v1.9/server-server-api/#get_matrixkeyv2server) already.

Reviewer 9: Sending `destination` is already required for all servers supporting Matrix v1.3 or higher, it's just receiving that's optional

> For compatibility with older servers, recipients should accept requests without this parameter, but MUST always send it.

Reviewer: > > we use REST principles in Matrix. I'm not sure this is the MSC where we want to really challenge that 😅
> 
> I don't really follow that this would violate REST principles. Do you mean that it's surprising for `GET /_matrix/federation/v1/media/download` to return different results depending on the server? We have that situation for [`GET /_matrix/key/v2/server`](https://spec.matrix.org/v1.9/server-server-api/#get_matrixkeyv2server) already.

I suppose I should have said *try* to follow REST principles 😭. REST generally tries to be semantically readable at the cost of verbosity, though it's certainly debatable if the server name is part of the resource ID (and thus should be part of the path). 

It sounds like folks are more in favour of dropping it though, so will do so. (As a process note: I'm not putting the MSC back into `needs-implementation` for this thread, though implementation will likely happen anyways).

token in the `Authorization` header.

The bridge might also choose to embed information such as the room that
referenced the media, and the time that the link was generated, in the URL.
Such mechanisms would allow the bridge to impose controls such as:

* Limiting the time a media link is valid for. Doing so would help prevent
  visibility to users who weren't participating in the chat.

* Rate-limiting the amount of media being shared in a particular room (in other
  words, avoiding the use of Matrix as a Warez distribution system).

#### Icons for "social login" flows

When a server supports multiple login providers, it provides the client with
icons for the login providers as `mxc:` media URIs. These must be accessible
without authentication (because the client has no access token at the time the
icons are displayed).

This remains a somewhat unsolved problem. Possibly the clients can continue
to call the legacy `/_matrix/media/v3/download` URI for now: ultimately this
problem will be solved by the transition to OIDC. Alternatively, we may need
to provide an alternative `/_matrix/client/v3/login/sso/icon/{idpId}` API

> specifically for access to these icon.

Reviewer 7: Another alternative is to allow https URLs there and have people host them outside of the Matrix media store.

   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}/{fileName}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaidfilename) | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}/{fileName}` | -                                                                 |
   | [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)          | `GET /_matrix/client/v1/media/thumbnail/{serverName}/{mediaId}`           | - |
   | [`GET /_matrix/media/v1/create`](https://spec.matrix.org/v1.10/client-server-api/#post_matrixmediav1create) | `GET /_matrix/client/v1/media/create` | - |

   **Note**: [`POST /_matrix/media/v3/upload`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixmediav3upload)
   is **not** modified or deprecated by this MSC: it is intended that it be brought into line with the other

>    endpoints by a future MSC, such as [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).

Reviewer 9: Wait, shouldn't `/create` and `/upload` be treated the same way (`/upload/{serverName}/{mediaID}` too)? i.e. neither should be modified by this MSC

@clokep said

> Or maybe this is not modified like /upload? But I don't see it mentioned at all.

in the previous thread, not sure if that was noticed

Reviewer: It was noticed, but MSC3911 wasn't really reviewed at the time. Having actually looked at it, it seems 3911 wants to take care of it - will move it there.

  words, avoiding the use of Matrix as a Warez distribution system).

#### Icons for "social login" flows

When a server supports multiple login providers, it provides the client with
icons for the login providers as `mxc:` media URIs. These must be accessible
without authentication (because the client has no access token at the time the
icons are displayed).

This remains a somewhat unsolved problem. Possibly the clients can continue
to call the legacy `/_matrix/media/v3/download` URI for now: ultimately this
problem will be solved by the transition to OIDC. Alternatively, a dedicated
API could be provided or permission to use HTTP(S) URLs to access the icons.
This support would come from a different MSC.

(This was previously discussed in
[MSC2858](https://github.com/matrix-org/matrix-spec-proposals/pull/2858#discussion_r543513811).)

## Potential issues

* Setting the `Authorization` header is particularly annoying for web clients.
  Service workers are seemingly the best option, though other options include
  locally-cached `blob:` URIs. Clients should note that caching media can lead
  to significant memory usage, particularly for large media. Service workers by

>   comparison allow for proxy-like behaviour.

Reviewer 11: This raises of question of, "why not support cookies", which could do with some discussion here.

  either. There are arguments for affecting both and either part equally. Typically,
  such a header would be used to resume failed downloads, though servers are
  already likely to discard received data and fail the associated client requests
  when the federation request fails. Therefore, servers are unlikely to use `Range`
  at all. As such, this proposal does not make a determination on how `Range`
  should be handled, and leaves it as an HTTP specification interpretation problem
  instead.

* [MSC3860](https://github.com/matrix-org/matrix-spec-proposals/pull/3860)-style
  redirects are harder to implement for the federation endpoints. It's presumed
  that CDNs can either cache the multipart type (later to be combined with linked
  media authentication, like [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911)),
  or the CDN can be somehow told the parameters it needs to return. For example,
  `Location: https://cdn.example.org/media?mx_json={}`. Popular CDN providers
  support this sort of request rewriting. Relatedly, [MSC4097](https://github.com/matrix-org/matrix-spec-proposals/pull/4097)
  may be of interest to readers.

> 

Reviewer 11: Is it worth mentioning how caching will be more complex?

Reviewer: ftr, I think this ties heavily into the thread regarding split endpoints - I'm considering it solved by that.

   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaid)            | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}`            | `GET /_matrix/federation/v1/media/download/{mediaId}`  |
   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}/{fileName}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaidfilename) | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}/{fileName}` | -                                                                 |
   | [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)          | `GET /_matrix/client/v1/media/thumbnail/{serverName}/{mediaId}`           | - |

   **Note**: [`POST /_matrix/media/v3/upload`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixmediav3upload)
   and [`GET /_matrix/media/v1/create`](https://spec.matrix.org/v1.10/client-server-api/#post_matrixmediav1create)

>    are **not** modified or deprecated by this MSC: it is intended that it be brought into line with the other

Reviewer 12: ```suggestion
   are **not** modified or deprecated by this MSC: it is intended that they be brought into line with the other
```

   endpoints by a future MSC, such as [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).

   **Note**: `/thumbnail` does not have a federation endpoint. It appears as though
   no servers request thumbnails over federation, and so it is not supported here.
   A later MSC may introduce such an endpoint.

   The new `/download` and `/thumbnail` endpoints additionally drop the `?allow_redirect`
   query parameters. Instead, the endpoints behave as though `allow_redirect=true` was
   set, regardless of actual value. See [this comment on MSC3860](https://github.com/matrix-org/matrix-spec-proposals/pull/3860/files#r1005176480)

>    for details.

Author: This seems to duplicate point 6?  (Making it a separate point seems clearer to me)

   After this proposal is released in a stable version of the specification, servers
   which support the new `download` and `thumbnail` endpoints SHOULD cease to serve

>    newly uploaded media from the unauthenticated versions. This includes media

Reviewer 10: Sorry, missed this got added.

If we do this won't this cause mass carnage in the ecosystem? Isn't this basically saying the entire client ecosystem needs to immediately support this MSC, or they probably will no longer be able to access newly uploaded media? 

Point 5 of this list has this as a `MAY` rather than `SHOULD` which feels more appropriate, though in general I think servers need to be reasonably conservative here unless they control which client their users use.

At some point we need to decide and announce timelines here, either as SCT, Synapse team or matrix.org operators.

Reviewer: good catch. This was added when the release was expected to be a month or so further out from FCP finishing. 

I'll move the language to be under point 5 instead, and find some words to make it clear that this should be done quickly (though not so quickly that everything breaks tomorrow sort of deal).

We plan to publish blog posts and social media under the matrix.org banner, as the Foundation.

   and
   [`/thumbnail`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)
   endpoints take an `allow_remote` query parameter, indicating whether the
   server should request remote media from other servers. This is redundant
   with the new endpoints, so will not be supported.

   Servers MUST NOT return remote media from `GET /_matrix/federation/v1/media/download`. The
   `serverName` is omitted from the endpoint's path to strongly enforce this - the `mediaId` in
   a request is assumed to be scoped to the target server.

   `/_matrix/client/v1/media/download` and
   `/_matrix/client/v1/media/thumbnail` return remote media as normal.

3. Authentication on all endpoints

   Currently, the `/download` and `/thumbnail` endpoints have no authentication
   requirements. Under this proposal, the new endpoints will be authenticated
   the same way as other endpoints: they will require an `Authorization` header
   which must be `Bearer {accessToken}` for `/_matrix/client`, or the signature
   for `/_matrix/federation`.

   The query string `?access_token` approach is not supported on the new endpoints,

>    as it is [deprecated](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/4126-deprecate-query-string-auth.md)

Author: it's easier for a reader to go from the PR to the doc than the other way around, so could we do:

```suggestion
   as it is [deprecated](https://github.com/matrix-org/matrix-spec-proposals/pull/4126)
```

Reviewer: The PR doesn't show any changes to accepted MSCs, unfortunately. Ideally this would be a link to the spec, but unstable has the same longevity problems.

   and [pending removal](https://github.com/matrix-org/matrix-spec-proposals/pull/4127).

>    See those MSCs for details.

Author: Much as I want to see the back of `?access_token`, I'm a little bit hesitant to make exceptions; imho either all endpoints should support it or none of them should. Otherwise the spec gets quite confusing.

Not a blocking comment though

Reviewer: I believe in practice all implementations accept the query string auth anyways, so maybe we allow it but place a warning in the spec that it's unsafe.

   **Note**: This fixes [matrix-spec#313](https://github.com/matrix-org/matrix-spec/issues/313).

4. Updated response format

   * For the new `/_matrix/client` endpoints, the response format is the same as
     the corresponding original endpoints.

   * To enable future expansion, for the new `/_matrix/federation` endpoints,
     the response is
     [`multipart/mixed`](https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html)
     content with exactly two parts: the first MUST be a JSON object (and should have a

>      `Content-type: application/json` header), and the second MUST be the media item
>      as per the original endpoints.

Reviewer 10: ```suggestion
     `Content-type: application/json` header), and the second MUST be the media item
     (OR include a `Location` header to redirect to the media item) as per the original
     endpoints.
```

Reviewer: I've clarified this using different language. Let me know if further clarifications are needed.

Reviewer: see 0b4f2c9d4d03752e8b57159f099712634427d862

     No properties are yet specified for the JSON object to be returned. One
     possible use is described by [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).

     An example response:

     ```
     Content-Type: multipart/mixed; boundary=gc0p4Jq0M2Yt08jU534c0p

     --gc0p4Jq0M2Yt08jU534c0p
     Content-Type: application/json

     {}

     --gc0p4Jq0M2Yt08jU534c0p
     Content-Type: text/plain

     This media is plain text. Maybe somebody used it as a paste bin.

     --gc0p4Jq0M2Yt08jU534c0p
     ```

     The second part (media item bytes) MAY include a [`Location` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location)
     to point to the raw media object instead of having bytes itself. Servers
     SHOULD NOT cache the `Location` header's value as the responding server may
     have applied time limits on its validity. Servers which don't immediately
     download the media from the provided URL should re-request the media and

>      metadata from the `/download` endpoint when ready for the media bytes.

Author: Presumably, in the presence of a `Location` header, the body is ignored, and SHOULD be empty?

     The `Location` header's URL does *not* require authentication, as it will
     typically be served by a CDN or other non-matrix server (thus being unable
     to verify any `X-Matrix` signatures, for example).

     An example response with a `Location` redirect would be:

     ```
     Content-Type: multipart/mixed; boundary=gc0p4Jq0M2Yt08jU534c0p

     --gc0p4Jq0M2Yt08jU534c0p
     Content-Type: application/json

     {}

     --gc0p4Jq0M2Yt08jU534c0p

>      Content-Type: text/plain

Author: what happens if this Content-Type is different from that on the target?

Seems like maybe the `Content-Type` MUST be omitted if there is a `Location`?

Reviewer 9: Yeah I think even dumb CDNs should always be capable of serving an accurate content type, so having it in the multipart response seems weird

Reviewer: This feels like a copy/paste fail. The content type on a redirect is unused.

>      Location: https://cdn.example.org/ab/c1/2345.txt

Reviewer 9: Not sure if redirects should have a content-type? Even dumb CDNs should serve accurate content types themselves.

Reviewer 9: Oops, Rich just commented the same 13 minutes ago

Author: [great minds think alike](https://github.com/matrix-org/matrix-spec-proposals/pull/3916#discussion_r1618434613)

referencing the media, include an HMAC to prevent tampering. For example:

```
https://<bridge_server>/media/{originServerName}/{mediaId}?mac={hmac}
```

When the bridge later receives a request to that URI, it checks the hmac,
and proxies the request to the homeserver, using its AS access
token in the `Authorization` header.

The bridge might also choose to embed information such as the room that
referenced the media, and the time that the link was generated, in the URL.
Such mechanisms would allow the bridge to impose controls such as:

* Limiting the time a media link is valid for. Doing so would help prevent
  visibility to users who weren't participating in the chat.

* Rate-limiting the amount of media being shared in a particular room (in other
  words, avoiding the use of Matrix as a Warez distribution system).

#### Icons for "social login" flows

When a server supports multiple login providers, it provides the client with
icons for the login providers as `mxc:` media URIs. These must be accessible
without authentication (because the client has no access token at the time the
icons are displayed).

> This remains a somewhat unsolved problem. Possibly the clients can continue

Author: Not a blocking comment, but: is it true that this remains unsolved? Sounds like you're gunning for a fairly aggressive timeline here - we had better figure out what we are doing about it?

Reviewer: yea, I think the best course of action here would be to allow these particular media items to bypass the freeze on the endpoints (which is a SHOULD/MAY anyways depending on which paragraph you read 😅). Will put that recommendation into the MSC.

  the homeserver's repository.

  The danger with this is that is that there's little stopping clients
  continuing to upload media as "public", negating all of the benefits in this
  MSC. It might be ok if media upload it was restricted to certain privileged
  users, or applied after the fact by a server administrator.

* We could simply require that `Authorization` headers be given when calling
  the existing endpoints. However, doing so would make it harder to evaluate
  the proportion of clients which have been updated, and it is a good
  opportunity to bring these endpoints into line with the rest of the
  client-server and federation APIs.

* There's no real need to rename `GET /_matrix/media/v3/preview_url` and `GET
  /_matrix/media/v3/config` at present, and we could just leave them in
  place. However, changing them at the same time makes the API more consistent.

  Conversely, we should make sure to rename `POST
  /_matrix/media/v3/upload` and `GET /_matrix/media/v3/create`. The reason to
  delay doing so is because MSC3911 will make more substantial changes to these
  endpoints, requiring another rename, and it is expected that both proposals

>   will be mergeed near to the same time as each other (so a double rename will

Author: ```suggestion
  will be merged near to the same time as each other (so a double rename will
```

  be confusing and unnecessary). However, if MSC3911 is delayed or rejected, we

>   should reconsider this.

Author: We *are* still planning to fast-follow with MSC3911, right?

Reviewer: Hopefully, yes. It won't be in the same spec release though.

   | [`GET /_matrix/media/v3/config`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3config) | `GET /_matrix/client/v1/media/config` | - |
   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaid)            | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}`            | `GET /_matrix/federation/v1/media/download/{mediaId}`  |
   | [`GET /_matrix/media/v3/download/{serverName}/{mediaId}/{fileName}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3downloadservernamemediaidfilename) | `GET /_matrix/client/v1/media/download/{serverName}/{mediaId}/{fileName}` | -                                                                 |
   | [`GET /_matrix/media/v3/thumbnail/{serverName}/{mediaId}`](https://spec.matrix.org/v1.6/client-server-api/#get_matrixmediav3thumbnailservernamemediaid)          | `GET /_matrix/client/v1/media/thumbnail/{serverName}/{mediaId}`           | - |

   **Note**: [`POST /_matrix/media/v3/upload`](https://spec.matrix.org/v1.6/client-server-api/#post_matrixmediav3upload)

>    and [`GET /_matrix/media/v1/create`](https://spec.matrix.org/v1.10/client-server-api/#post_matrixmediav1create)

Reviewer 7: ```suggestion
   and [`POST /_matrix/media/v1/create`](https://spec.matrix.org/v1.10/client-server-api/#post_matrixmediav1create)
```

this proposal ("Backwards compatibility mechanisms"). Clients SHOULD continue to
use the unauthenticated media download/thumbnail endpoints to access these icons
until a future MSC can improve the situation.

The proposal author expects that the spec's transition to OIDC will ultimately
replace this mechanism, or that an MSC could be opened to allow HTTP(S) URLs in
place of `mxc:` URIs.

(This was previously discussed in
[MSC2858](https://github.com/matrix-org/matrix-spec-proposals/pull/2858#discussion_r543513811).)

## Potential issues

* Setting the `Authorization` header is particularly annoying for web clients.
  Service workers are seemingly the best option, though other options include
  locally-cached `blob:` URIs. Clients should note that caching media can lead
  to significant memory usage, particularly for large media. Service workers by
  comparison allow for proxy-like behaviour.

  Cookies are a plausible mechanism for sharing session information between
  requests without having to set headers, though would be a relatively bespoke
  authentication method for Matrix. Additionally, many Matrix users have cookies

>   disabled due to the advertising and tracking use cases common across the web.

Reviewer: Not hard data, but for the last 6 years it's felt like every third user for Element Web has disabled these things and more :p

>   disabled due to the advertising and tracking use cases common across the web.

Reviewer 11: Mmm, the latter seems spurious since cookies and localstorage will almost always be enabled/disabled as a pair, and Element has been relying on localstorage for years. The former is true in that setting a cookie for the homeserver would be a little weird. In my mind, the pros / cons are:

 * Cookie would make all media requests magically work on that browser session, even if the user opened something (non-encrypted) in a tab manually.
 * The above could cause them to think the link will work if sent to someone else, which it won't.
 * It would be a slightly odd way of using cookies since this is a federated world and the HS is not necessarily the same endpoint as your client.
 * It would cause every other C/S API request to also get the cookie sent, confusing auth.
 * Simpler if all C/S API auth works the same way.

I'm not going to make this a concern as I don't think it's blocking, but it might be worth a quick sanity check to make sure we really are committing to the right thing.

   is **not** modified by this MSC: it is intended that it be brought into line with the other
   endpoints by a future MSC, such as [MSC3911](https://github.com/matrix-org/matrix-spec-proposals/pull/3911).

   After this proposal is released in a stable version of the specification, servers
   which support the new `download` and `thumbnail` endpoints should cease to serve
   newly uploaded media from the unauthenticated versions. This includes media
   uploaded by local users and requests for not-yet-cached remote media. This is

>    done with a 404 `M_NOT_FOUND` error, as though the media doesn't exist.

Reviewer 13: I think unauthorized with a redirect to the new endpoints is more semantically correct and self-documenting.

Reviewer: All HTTP redirects just send the exact same headers, so it'd be a 404, 403, or 401 anyways. We also can't redirect unless the requester indicates they support redirects, which increases complexity.

Given the constraints, a 404 should be clear enough imo.

Reviewer: (resolving - incorporated into the MSC, I believe. Please re-raise if not)
